stablehlo/reference/Element.cpp

```diff
@@ -362,44 +366,124 @@ Element Element::operator/(const Element &other) const {
                                      debugString(type).c_str()));
 }
 
-Element Element::operator*(const Element &other) const {
-  return map(
-      *this, other, [](APInt lhs, APInt rhs) { return lhs * rhs; },
-      [](bool lhs, bool rhs) -> bool { return lhs & rhs; },
-      [](APFloat lhs, APFloat rhs) { return lhs * rhs; },
-      [](std::complex<APFloat> lhs, std::complex<APFloat> rhs) {
-        // TODO(#226): Use std::complex::operator*
-        auto resultReal = lhs.real() * rhs.real() - lhs.imag() * rhs.imag();
-        auto resultImag = lhs.real() * rhs.imag() + lhs.imag() * rhs.real();
-        return std::complex<APFloat>(resultReal, resultImag);
-      });
+bool Element::operator<(const Element &other) const {
+  Type type = other.getType();
+  if (type_ != type)
+    report_fatal_error(invalidArgument("Element types don't match: %s vs %s",
+                                       debugString(type_).c_str(),
+                                       debugString(type).c_str()));
+
+  if (isSupportedIntegerType(type)) {
+    auto intLhs = getIntegerValue();
+    auto intRhs = other.getIntegerValue();
+    return isSupportedSignedIntegerType(type) ? intLhs.slt(intRhs)
+                                              : intLhs.ult(intRhs);
+  }
+
+  if (isSupportedBooleanType(type)) {
+    auto boolLhs = getBooleanValue();
+    auto boolRhs = other.getBooleanValue();
+    return boolLhs < boolRhs;
+  }
+
+  if (isSupportedFloatType(type)) {
+    auto floatLhs = getFloatValue();
+    auto floatRhs = other.getFloatValue();
+    return floatLhs < floatRhs;
+  }
+
+  if (isSupportedComplexType(type)) {
+    auto complexLhs = getComplexValue();
+    auto complexRhs = other.getComplexValue();
+    if (complexLhs.real() < complexRhs.real()) return true;
+    if (!(complexLhs.real() == complexRhs.real())) return false;
+    return complexLhs.imag() < complexRhs.imag();
+  }
+
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
 }
```

Comment:
The current `map` function assumes the element types do not change, but regardless of the input element type, these functions return a `bool`, and the current map implementations do not allow specifying the input type to check and the result type to return. Also, `map` does not differentiate signed vs unsigned integers, so it will require some refactoring to use it. I've opened a ticket for this #1303.